src/category_theory/limits/shapes/cartesian_closed.lean

```diff
@@ -0,0 +1,387 @@
+/-
+Copyright (c) 2020 Bhavik Mehta, Edward Ayers, Thomas Read. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta, Edward Ayers, Thomas Read
+-/
+
+import category_theory.limits.shapes.binary_products
+import category_theory.limits.shapes.constructions.preserve_binary_products
+import category_theory.adjunction
+import category_theory.epi_mono
+
+/-!
+# Cartesian closed categories
+
+Define exponentiable objects and cartesian closed categories.
+Show that exponential forms a difunctor.
+Define the exponential comparison morphisms.
+-/
+universes v u u₂
+
+namespace category_theory
+
+open limits category
+
+/-- An object `X` is *exponentiable* if `(X × -)` is a left adjoint. -/
+class exponentiable {C : Type u} [category.{v} C] [has_finite_products.{v} C] (X : C) :=
+(is_adj : is_left_adjoint (prod_functor.obj X))
+
+/--
+If `X` and `Y` are exponentiable then `X ⨯ Y` is.
+This isn't an instance because it's not usually how we want to construct exponentials
+-/
+def binary_product_exponentiable {C : Type u} [category.{v} C] [has_finite_products.{v} C] {X Y : C}
+  (hX : exponentiable X) (hY : exponentiable Y) : exponentiable (X ⨯ Y) :=
+{ is_adj :=
+  by haveI := hX.is_adj; haveI := hY.is_adj;
```

Comment:
Only use `;` in `by` if there are actually multiple goals. Otherwise use `by { X, Y }` with braces.

Also, use `begin ... end` in favour of multi-line `by` statements.